lambeq/training/nelder_mead_optimizer.py

```diff
@@ -0,0 +1,394 @@
+                    'lower bounds must be less than upper bounds.'
+                )
+            if np.any(lb > self.model.weights) or np.any(
+                self.model.weights > ub
+            ):
+                warnings.warn(
+                    'Initial value of model weights is not within the bounds.',
+                    stacklevel=2
+                )
+
+            self.project = lambda x: x.clip(bds[:, 0], bds[:, 1])
+            self.model.weights = self.project(self.model.weights)
+
+        N = len(self.model.weights)
+        if self.initial_simplex is None:
+            self.sim = np.empty((N + 1, N), dtype=self.model.weights.dtype)
+            self.sim[0] = model.weights
+            for k in range(N):
+                y = np.array(self.model.weights, copy=True)
+                if y[k] != 0:
+                    y[k] = (1 + self.nonzdelt) * y[k]
+                else:
+                    y[k] = self.zdelt
+                self.sim[k + 1] = y
+        else:
+            self.sim = np.asfarray(self.initial_simplex).copy()
+            if (
+                self.sim.ndim != 2
+                or self.sim.shape[0] != self.sim.shape[1] + 1
+            ):
+                raise ValueError(
+                    '`initial_simplex` should be an array of shape'
+                    f'({N+1},{N})'
+                )
+            if len(self.model.weights) != self.sim.shape[1]:
+                raise ValueError(
+                    'Size of `initial_simplex` is not consistent with'
+                    '`model.weights`'
+                )
+            N = self.sim.shape[1]
+
+        self.sim = self.project(self.sim)
+        self.one2np1 = list(range(1, N + 1))
+        self.fsim = np.full((N + 1,), np.inf, dtype=float)
+        self.N = N
+        self.first_iter = True  # flag for first iteration
+
+    def backward(self, batch: tuple[Iterable[Any], np.ndarray]) -> float:
+        """Calculate the gradients of the loss function.
+
+        The gradients are calculated with respect to the model
+        parameters.
+
+        Parameters
+        ----------
+        batch : tuple of Iterable and numpy.ndarray
+            Current batch. Contains an Iterable of diagrams in index 0,
+            and the targets in index 1.
+
+        Returns
+        -------
+        float
+            The calculated loss.
+
+        """
+
+        diagrams, targets = batch
+        diags_gen = flatten(diagrams)
+
+        # the symbolic parameters
+        parameters = self.model.symbols
+
+        relevant_params = set.union(
+            *[diag.free_symbols for diag in diags_gen]
+        )
+        mask = np.array(
+            [1 if sym in relevant_params else 0 for sym in parameters]
+        )
+
+        # Initialize ind, sim and fsim
+        if self.first_iter:
+            try:
+                for k in range(self.N + 1):
+                    self.fsim[k] = self.func(diagrams, targets, self.sim[k])
+            except BaseException:
```

Comment:
Please make this catch a more specific exception, if you can. At the very least, make it `Exception`. Using `BaseException` is dangerous because it catches errors like `KeyboardInterrupt`, making it impossible to force quit the program if needed.

Reply:
Shall I consider this resolved?